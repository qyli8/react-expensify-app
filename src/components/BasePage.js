import React from 'react';
import NotFoundPage from '../components/NotFoundPage'

// nodejs library that concatenates classes
import classNames from "classnames";
// @material-ui/core components
import { makeStyles } from "@material-ui/core/styles";
// @material-ui/icons
// core components
import Footer from "../losecomponents/Footer/Footer.js";
import GridContainer from "../losecomponents/Grid/GridContainer.js";
import GridItem from "../losecomponents/Grid/GridItem.js";
import Parallax from "../losecomponents/Parallax/Parallax.js";
import Header from "../losecomponents/Header/Header.js";
import HeaderLinks from "../losecomponents/Header/HeaderLinks.js";
import styles from "../assets/jss/material-kit-react/views/components.js";

const useStyles = makeStyles(styles);
const BasePage = (props) => {
  console.log('basepage props: ', props)
  const classes = useStyles();
  // const { ...rest } = props;
  const Component = props.component
  const { ...rest } = props;
 
  return (
      <div>
          <Header
        absolute
        color="transparent"
        rightLinks={<HeaderLinks />}
        {...rest}
      />        
        <Parallax small image="../assets/img/lp.jpg">
          <div className={classes.container}>
            <GridContainer>
              <GridItem>
                <div className={classes.brand}>
                  {props.title && <h1 className={classes.title}>{props.title}</h1>}
                </div>
              </GridItem>
            </GridContainer>
          </div>
        </Parallax>
        <div className={classNames(classes.main, classes.mainRaised)}>
          <div className={classes.container} style={{ paddingBottom: "2rem" }}>
            <Component {...props}/>
          </div>
        </div>
        <Footer />
      </div>

  )
}
export default BasePage;
