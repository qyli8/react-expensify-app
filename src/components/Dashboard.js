import React from "react";
import ExpenseSumary from './ExpenseSummary'
import { connect } from 'react-redux';
import moment from 'moment'
import {
  XAxis,
  YAxis,
  BarChart,
  CartesianGrid,
  Tooltip,
  Legend,
  Bar
} from 'recharts'

export const DashBoardPage = (props) => {
  const data = [
    {
      "name": moment().subtract(7, 'months').format('MMM YYYY'),
      "expense": props.monthSevenExpenses
    },
    {
      "name": moment().subtract(6, 'months').format('MMM YYYY'),
      "expense": props.monthSixExpenses
    },
    {
      "name": moment().subtract(5, 'months').format('MMM YYYY'),
      "expense": props.monthFiveExpenses,
    },
    {
      "name": moment().subtract(4, 'months').format('MMM YYYY'),
      "expense": props.monthFourExpenses,
    },
    {
      "name": moment().subtract(3, 'months').format('MMM YYYY'),
      "expense": props.monthThreeExpenses,
    },
    {
      "name": moment().subtract(2, 'months').format('MMM YYYY'),
      "expense": props.monthTwoExpenses,
    },
    {
      "name": moment().subtract(1, 'months').format('MMM YYYY'),
      "expense": props.monthOneExpenses,
    }
  ]
  return (
    <div>
      <ExpenseSumary path={props.match.path} />
      <div className="data_row">
        <div className="data_title">
          <BarChart width={500} height={300} data={data}>
            <CartesianGrid strokeDasharray="3 3" />
            <XAxis dataKey="name" />
            <YAxis />
            <Tooltip />
            <Legend />
            <Bar dataKey="expense" fill="#8884d8" />
            {/* <Bar dataKey="uv" fill="#82ca9d" /> */}
          </BarChart>
        </div>
      </div>
    </div>
  );
}



const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
    monthOneExpenses: state.expenses.filter((e) => moment(e.createdAt).get('month') === moment().subtract(1, 'months').get('month')).reduce((total, currentExpense) => total + parseFloat(currentExpense.amount), 0),
    monthTwoExpenses: state.expenses.filter((e) => moment(e.createdAt).get('month') === moment().subtract(2, 'months').get('month')).reduce((total, currentExpense) => total + parseFloat(currentExpense.amount), 0),
    monthThreeExpenses: state.expenses.filter((e) => moment(e.createdAt).get('month') === moment().subtract(3, 'months').get('month')).reduce((total, currentExpense) => total + parseFloat(currentExpense.amount), 0),
    monthFourExpenses: state.expenses.filter((e) => moment(e.createdAt).get('month') === moment().subtract(4, 'months').get('month')).reduce((total, currentExpense) => total + parseFloat(currentExpense.amount), 0),
    monthFiveExpenses: state.expenses.filter((e) => moment(e.createdAt).get('month') === moment().subtract(5, 'months').get('month')).reduce((total, currentExpense) => total + parseFloat(currentExpense.amount), 0),
    monthSixExpenses: state.expenses.filter((e) => moment(e.createdAt).get('month') === moment().subtract(6, 'months').get('month')).reduce((total, currentExpense) => total + parseFloat(currentExpense.amount), 0),
    monthSevenExpenses: state.expenses.filter((e) => moment(e.createdAt).get('month') === moment().subtract(7, 'months').get('month')).reduce((total, currentExpense) => total + parseFloat(currentExpense.amount), 0),
    last7Months: [
      moment().subtract(1, 'months').format('MMM', 'YYYY'),
      moment().subtract(2, 'months').format('MMM', 'YYYY'),
      moment().subtract(3, 'months').format('MMM', 'YYYY'),
      moment().subtract(4, 'months').format('MMM', 'YYYY'),
      moment().subtract(5, 'months').format('MMM', 'YYYY'),
      moment().subtract(6, 'months').format('MMM', 'YYYY'),
      moment().subtract(7, 'months').format('MMM', 'YYYY')
    ]
  }
}

export default connect(mapStateToProps)(DashBoardPage);