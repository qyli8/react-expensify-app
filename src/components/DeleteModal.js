
import React from 'react';
import Modal from '@material-ui/core/Modal'
import Backdrop from '@material-ui/core/Backdrop';
import Fade from '@material-ui/core/Fade'
import Fab from '@material-ui/core/Fab';
import { connect } from 'react-redux'
import { ModalClose} from '../actions/modal'
import { withStyles, makeStyles } from '@material-ui/core/styles';

const PrimaryButton = withStyles(() => ({
  "extended": {
    width: "5rem",
    fontSize: "13px",
    margin: "2% 0 0 0"
  }
}))(Fab)

const useStyles = makeStyles(theme => ({
  modal: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  paper: {
    backgroundColor: "#f7f7ff",
    border: '2px solid #000',
    boxShadow: theme.shadows[5],
    padding: theme.spacing(2, 4, 3),
  },
  modle_options:{
    display:'flex',
    alignItems: 'center',
    justifyContent: 'space-around',
    fontSize: '15px'
  },
  modle_message:{
    fontSize: '17px'
  }
}));

const DeleteModal = ({open, closeModal, description, removeExpense}) =>{ 
  const classes=useStyles()
  return(
  <Modal
    aria-labelledby="transition-modal-title"
    aria-describedby="transition-modal-description"
    className={classes.modal}
    open={open}
    onClose={closeModal}
    closeAfterTransition
    BackdropComponent={Backdrop}
    BackdropProps={{
      timeout: 500,
    }}
  >
    <Fade in={open}>
      <div className={classes.paper}>
        <p  className={classes["modle_message"]}>Remove {description}, are you sure?</p>
        <div className={classes.modle_options}>
          <PrimaryButton onClick={removeExpense} variant="extended" color="primary" >
          YES
        </PrimaryButton>
          <PrimaryButton onClick={closeModal} variant="extended" color="secondary">
            NO
        </PrimaryButton>
        </div>
      </div>
    </Fade>
  </Modal>

)}

const mapStateToProps = (state, props) => {
  return {
    open: state.modal.open
  }
}

const matchDispatchaToProps = (dispatch) => ({
  closeModal:()=>(dispatch(ModalClose()))
})
export default connect(mapStateToProps, matchDispatchaToProps)(DeleteModal)