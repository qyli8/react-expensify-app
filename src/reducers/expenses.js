//Expenses Reducer
const expensesReducerDefaultState=[]

const expensesReducer = (state=expensesReducerDefaultState, action) =>{
  switch(action.type){
    case 'ADD_EXPENSE':
    // spread operator allow concatination and create new array
     return [...state, action.expense]
    case 'REMOVE_EXPENSE':
      return state.filter(i=>i.id.localeCompare(action.id)!=0)
    case 'EDIT_EXPENSE':
      return state.map( (expense)=>{
        if(expense.id === action.id){
          return ({
            ... expense,
            ...action.updates
          })
        }
        else{
          return expense
        }
      })
    default:
    return state
  }
}

export default expensesReducer;