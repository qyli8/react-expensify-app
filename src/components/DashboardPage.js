import React from 'react'
import DashBoard from "./Dashboard";
import BasePage from "./BasePage"

export const DashBoardPage = (props) => {
  console.log('edit page props: ' , props)
  return(
  <BasePage title="WELCOME Joanna Li" component={DashBoard} {...props}/>
)}

export default DashBoardPage

