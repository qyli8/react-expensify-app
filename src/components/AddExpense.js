import React from "react";
import ExpenseForm from './ExpressForm'
import { connect } from "react-redux";
import { addExpense } from '../actions/expenses'
import { home } from '../actions/path'
import Paper from '@material-ui/core/Paper';
import { withStyles } from '@material-ui/core/styles';
import Fab from '@material-ui/core/Fab';

const CustomPaper = withStyles(() => ({
  root: { padding: "30px" }
}))(Paper)
const PrimaryButton = withStyles(() => ({
  "extended": {
    width: "15rem",
    fontSize: "13px",
    margin: "2% 0 0 0"
  }
}))(Fab)

export class AddExpense extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/home')
    this.props.home('HOME')
  }
  cancel = () => {
    this.props.home('HOME')
    this.props.history.push("/home")
  }

  render() {
    return (
      <div className="page-form_padding">
        <div>
        <CustomPaper>
          <ExpenseForm
            onSubmit={this.onSubmit}
          />
          <PrimaryButton onClick={this.cancel} variant="extended" color="primary" aria-label="cancel"   >
            Cancel
        </PrimaryButton>
        </CustomPaper>
      </div>
      </div>
    )
  }
}

const matchDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense)),
  home: (path) => dispatch(home(path)),
})

export default connect(undefined, matchDispatchToProps)(AddExpense)