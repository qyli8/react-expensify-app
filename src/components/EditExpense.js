import React from "react";
import { connect } from 'react-redux'
import ExpenseForm from './ExpressForm'
import { editExpense, removeExpense } from '../actions/expenses'
import Paper from '@material-ui/core/Paper';
import { withStyles} from '@material-ui/core/styles';

const CustomPaper = withStyles(()=>({
  root:{padding:"10px"}
}))(Paper)

export class EditExpensePage extends React.Component {
  editExpense = (expense) => {

    this.props.editExpense(this.props.match.params.id, expense)
    this.props.history.push("/home")
  }
  removeExpense = (e) => {
    this.props.removeExpense(this.props.expense)
    this.props.history.push("/home")
  }

  cancel=()=>{
    this.props.history.push("/home")
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Edit Expense</h2>
          </div>
        </div>
        <div className="content-container">
          <div className="edit-form-container">
          <CustomPaper >
          <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.editExpense}
        />
        <div className="button_group">
        <button className="button-sytle__cancel" onClick={this.cancel}>
          Cancel
        </button>
        <button className="button-sytle__secondary" onClick={this.removeExpense}>
          Remove
        </button>
        </div>
        </CustomPaper>
          </div>
        
        </div>
        
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
}

const matchDispatchaToProps = (dispatch) => ({
  removeExpense: (expense) => dispatch(removeExpense(expense)),
  editExpense: (id, expense) => dispatch(editExpense(id, expense))
})

export default connect(mapStateToProps, matchDispatchaToProps)(EditExpensePage);