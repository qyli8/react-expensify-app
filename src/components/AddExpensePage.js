import React from 'react'
import AddExpense from "./AddExpense";
import BasePage from "./BasePage"

export const AddExpensePage = (props) => {
  console.log('edit page props: ' , props)
  return(
  <BasePage title="Add Expense" component={AddExpense} {...props}/>
)}

export default AddExpensePage