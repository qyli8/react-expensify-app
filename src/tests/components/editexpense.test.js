import React from 'react'
import {shallow} from 'enzyme'
import expenses from '../fixture/expenses'
import {EditExpensePage} from '../../components/EditExpense'
import moment from 'moment'

let editExpense, removeExpense, wrapper, history
beforeEach(()=>{
  editExpense = jest.fn()
  removeExpense = jest.fn()
  history = {push: jest.fn()}
  const id=expenses[1].id
  wrapper = shallow(<EditExpensePage 
    expense={expenses[1]}
    editExpense={editExpense}
    removeExpense={removeExpense}
    match={{params:{id}}}
    history={history}
    />)
})

test('should render edit expense',()=>{
  expect(wrapper).toMatchSnapshot()
})

test('should handle EditExpense',()=>{
  const editedExpense={
    note:'abc',
    amount:'50671111',
    createdAt:moment(0).subtract(10, 'days').valueOf()
  }
  wrapper.find('ExpenseForm').prop('onSubmit')(editedExpense)
  expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id,editedExpense)
  expect(history.push).toHaveBeenLastCalledWith("/")
})

test('should handle remove expense',()=>{
  wrapper.find('button').simulate('click')
  expect(removeExpense).toHaveBeenLastCalledWith(expenses[1])
  expect(history.push).toHaveBeenLastCalledWith("/")
})