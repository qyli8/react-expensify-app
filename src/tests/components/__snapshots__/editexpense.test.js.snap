// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title"
      >
        Edit Expense
      </h2>
    </div>
  </div>
  <div
    className="content-container"
  >
    <div
      className="edit-form-container"
    >
      <ExpenseForm
        expense={
          Object {
            "amount": "5067000",
            "createdAt": -345600000,
            "description": "item 2",
            "id": "2",
            "note": "",
          }
        }
        onSubmit={[Function]}
      />
      <button
        className="button-sytle__secondary"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
  </div>
</div>
`;
